File: rsleetcode/src/easy/symmetric_tree.rs
//! 101. Symmetric Tree
//!
//! https://leetcode.com/problems/symmetric-tree/

use std::cell::RefCell;
use std::rc::Rc;
type Tree = Rc<RefCell<TreeNode>>;
type OptionTree = Option<Tree>;

// Definition for a binary tree node.
#[derive(Debug, PartialEq, Eq)]
pub struct TreeNode {
    pub val: i32,
    pub left: OptionTree,
    pub right: OptionTree,
}

impl TreeNode {
    #[inline]
    pub fn new(val: i32) -> Self {
        TreeNode {
            val,
            left: None,
            right: None,
        }
    }
}

pub struct Solution {}

impl Solution {
    pub fn is_symmetric(root: OptionTree) -> bool {
        fn compare(left: OptionTree, right: OptionTree) -> bool {
            if left.is_none() && right.is_none() {
                return true
            };

            if let (Some(l_node), Some(r_node)) = (left, right) {
                if l_node.borrow().val != r_node.borrow().val {
                    return false
                };
                compare(l_node.borrow().left.clone(), r_node.borrow().right.clone())
                && compare(r_node.borrow().left.clone(), l_node.borrow().right.clone())
            } else {
                false
            }
        }

        let root = root.unwrap();
        let res = compare(root.borrow().left.clone(), root.borrow().right.clone());
        res
    }
}

#[cfg(test)]
mod tests {
    use super::*;

    fn create_tree(vals: &Vec<Option<i32>>) -> OptionTree {
        if vals.is_empty() {
            return None;
        }

        fn create(i: usize, vals: &Vec<Option<i32>>) -> OptionTree {
            if i >= vals.len() || vals[i].is_none() {
                return None;
            }
            let mut node = TreeNode::new(vals[i].unwrap());
            node.left = create(i * 2 + 1, vals);
            node.right = create(i * 2 + 2, vals);
            Some(Rc::new(RefCell::new(node)))
        }

        create(0, &vals)
    }

    #[test]
    fn test_1() {
        let vals = vec![
            Some(1),
            Some(2),
            Some(2),
            Some(3),
            Some(4),
            Some(4),
            Some(3),
        ];
        let root = create_tree(&vals);

        assert_eq!(Solution::is_symmetric(root), true);
    }

    #[test]
    fn test_2() {
        let vals = vec![Some(1), Some(2), Some(2), None, Some(3), None, Some(3)];
        let root = create_tree(&vals);

        assert_eq!(Solution::is_symmetric(root), false);
    }
}

File: pyleetcode/pyleetcode/easy/symmetric_tree.py
"""101. Symmetric Tree

https://leetcode.com/problems/symmetric-tree/
"""
from __future__ import annotations
from typing import Optional, Union, Any  # noqa: F401


class TreeNode:

    def __init__(self, val=0, left=None, right=None):
        self.val = val
        self.left = left
        self.right = right

    @classmethod
    def new(cls, vals: list[Union[int, None]]) -> Optional[TreeNode]:
        if len(vals) == 0:
            return None

        def new(i: int) -> Optional[TreeNode]:
            if i >= len(vals) or vals[i] is None:
                return None
            node = TreeNode(vals[i])
            node.left = new(i * 2 + 1)
            node.right = new(i * 2 + 2)
            return node

        return new(0)

    def __str__(self) -> str:
        return str(self.val)

    def __repr__(self) -> str:
        return str(self)


class Solution:

    def isSymmetric(self, root: Optional[TreeNode]) -> bool:  # noqa: N802
        l_stack: list[TreeNode] = [root.left]
        r_stack: list[TreeNode] = [root.right]
        while l_stack and r_stack:
            l_node = l_stack.pop()
            r_node = r_stack.pop()
            if l_node is None and r_node is None:
                continue

            if l_node is None and r_node is not None:
                return False
            if l_node is not None and r_node is None:
                return False

            if l_node.val != r_node.val:
                return False
            l_stack.append(l_node.right)
            l_stack.append(l_node.left)
            r_stack.append(r_node.left)
            r_stack.append(r_node.right)

        return True

    def process(self, *args, **kwargs) -> bool:  # noqa: N802
        return self.isSymmetric(*args, **kwargs)
